loky/backend/_thread_pool_limiters.py

```diff
@@ -0,0 +1,492 @@
+UINT_HALF_SYSTEM = ctypes.c_uint32 if sys.maxsize > 2**32 else ctypes.c_uint16
+
+
+class dl_phdr_info(ctypes.Structure):
+    _fields_ = [
+        ("dlpi_addr",  UINT_SYSTEM),       # Base address of object
+        ("dlpi_name",  ctypes.c_char_p),   # path to the library
+        ("dlpi_phdr",  ctypes.c_void_p),   # pointer on dlpi_headers
+        ("dlpi_phnum",  UINT_HALF_SYSTEM)  # number of element in dlpi_phdr
+        ]
+
+
+# map a library (openmp, openblas, mkl) to set and get functions
+LIB_TO_FUNC_MAP = {
+    "openmp": {
+        "set": "omp_set_num_threads",
+        "get": "omp_get_max_threads"},
+    "openblas": {
+        "set": "openblas_set_num_threads",
+        "get": "openblas_get_num_threads"},
+    "mkl": {
+        "set": "MKL_Set_Num_Threads",
+        "get": "MKL_Get_Max_Threads"},
+}
+
+
+# Supported implementation, indexed with their name. The items hold the name of
+# the implementation, the prefix of the name of the loaded library and the name
+# of the library to call, matching the LIB_TO_FUNC_MAP keys.
+SUPPORTED_IMPLEMENTATION = {
+    "openmp_intel": {
+        "filename_prefixes": ("libiomp",),
+        "library": "openmp"
+    },
+    "openmp_gnu": {
+        "filename_prefixes": ("libgomp",),
+        "library": "openmp"
+    },
+    "openmp_llvm": {
+        "filename_prefixes": ("libomp",),
+        "library": "openmp"
+    },
+    "openmp_msvc": {
+        "filename_prefixes": ("vcomp",),
+        "library": "openmp"
+    },
+    "openblas": {
+        "filename_prefixes": ("libopenblas",),
+        "library": "openblas"
+    },
+    "mkl": {
+        "filename_prefixes": ("libmkl_rt", "mkl_rt",),
+        "library": "mkl"
+    },
+}
+
+
+class _ThreadPoolLibrariesWrapper:
+    # Wrapper around classic C-libraries for scientific computations which use
+    # thread-pools. Give access to functions to set and get the maximum number
+    # of threads they are allowed to used for inner parallelism.
+
+    cls_thread_locals = threading.local()
+
+    def __init__(self):
+        self._load()
+
+    def _load(self):
+        if sys.platform == "darwin":
+            self._modules = self._find_modules_with_clibs_dyld()
+        elif sys.platform == "win32":
+            self._modules = self._find_modules_with_enum_process_module_ex()
+        else:
+            self._modules = self._find_modules_with_clibs_dl_iterate_phdr()
+
+    def _unload(self):
+        del self._modules
+
+    def starts_with_any(self, library_basename, filename_prefixes):
+        return any([library_basename.startswith(prefix)
+                    for prefix in filename_prefixes])
+
+    def _is_supported_implementation(self, module_path):
+        module_name = os.path.basename(module_path).lower()
+        for name, info in SUPPORTED_IMPLEMENTATION.items():
+            if self.starts_with_any(module_name, info['filename_prefixes']):
+                return name, info['library']
+        return None, None
```

Comment:
Should we keep the different names `openmp_msvc`, `openmp_gnu`,... or merge them as `openmp` with multiple `filename_prefix`?

I think the later makes more sense with our current implementation.

Reply:
I thought about that. I am fine with both as long as we are not able to retrieve the version number.

Reply:
Seems good. In case of nested parallelism BLAS inside prange, setting the number of threads for the inner BLAS can be done through the "blas" user-api, even if BLAS is linked to OpenMP. So there should never be a case where we need to explicitly restrict one OpenMP and not the other.

>  I am fine with both as long as we are not able to retrieve the version number.

OpenMP does not exposes it's version so I would not worry about that.

Reply:
OpenMP no but the individual openmp runtime library might have a version introspection API.

Reply:
None of them does. Retrieving the version is a real pain :)
What you have to do is use the `_OPENMP` preprocessor macro in a program to get it's value. This value is a date, not a version number. Finally you have to use a map date/version to match the version from the date...